Add three examples (#513)
lethe-cfd/lethe: doc/source/examples/multiphysics/rayleigh-benard-convection/rayleigh-benard-convection.rst

.. image:: images/geometry.png
    :alt: Schematic
    :align: center
    :width: 400

The incompressible Navier-Stokes equations with a Boussinesq approximation for the buoyant force are:
    .. math::
        \nabla \cdot {\bf{u}} = 0

    .. math::
        \rho \frac{\partial {\bf{u}}}{\partial t} + \rho ({\bf{u}} \cdot \nabla) {\bf{u}} = -\nabla p + \nabla \cdot {\bf{\tau}} - \rho \beta {\bf{g}} (T - T_0)

where :math:`\beta` and :math:`T_0` denote thermal expansion coefficient and a reference temperature, respectively.

A two-dimensional block of fluid is heated from its bottom wall at :math:`t = 0` s. The temperature of the bottom wall is equal to math:`T_h=50`, the temperature of the top wall is equal to math:`T_c=0`, and the left and right walls are insulated. By heating the fluid from the bottom wall, the buoyant force (natural convection) creates vortices inside the fluid. The shape and number of these vortices mainly depend on the Rayleigh number [1, 2]:

    .. math::

>         \text{Ra} = \frac{\rho^2 \beta g (T_h - T_c) H^3}{k \mu}

Reviewer: ```suggestion
        \text{Ra} = \frac{\rho^2 \beta g (T_h - T_c) H^3 c_p }{k \mu}
```
I think there is a cp missing?

Author: Absolutely true, thank you!

>         \text{Ra} = \frac{\rho^2 \beta g (T_h - T_c) H^3}{k \mu}

Author: I've used tab in all the examples that I've added. I will change them all in another PR.

> where :math:`\rho` is the fluid density, :math:`\beta` denotes the thermal expansion coefficient, :math:`g` is the magnitude of gravitational acceleration, :math:`T_h` and :math:`T_c` are the hot wall and the cold wall temperatures, :math:`H` denotes the characteristic length, :math:`k` is the thermal conduction coefficient, and :math:`\mu` and :math:`\nu` are the dynamic and kinematic viscosities, :math:`c_p` is the specific thermal capacity, and :math:`\lambda` is the enthalpy of melting.

Reviewer: beta, T_c and T_h are already defined above. 

```suggestion
where :math:`\rho` is the fluid density, :math:`g` is the magnitude of gravitational acceleration, :math:`H` denotes the characteristic length, :math:`k` is the thermal conduction coefficient, and :math:`\mu` and :math:`\nu` are the dynamic and kinematic viscosities, and :math:`c_p` is the specific thermal capacity,.
```

> In this example, we simulate the Rayleigh-Benard convection problem at two Rayleigh numbers of 10000 and 25000. According to the literature [1, 2], we should see different numbers (2 and 3 vortices at Ra=10000 and 25000, respectively) of vortices in the fluid at these two Rayleigh numbers. Note that in this example, we change the Ra number by changing the thermal conduction coefficient of the fluid.

Reviewer 2: ```suggestion
In this example, we simulate the Rayleigh-Bénard convection problem at two Rayleigh numbers of 10000 and 25000. According to the literature [1, 2], we should see different numbers (2 and 3 vortices at Ra=10000 and 25000, respectively) of vortices in the fluid at these two Rayleigh numbers. Note that in this example, we change the Ra number by changing the thermal conduction coefficient of the fluid.
```

> In this example, we simulate the Rayleigh-Benard convection problem at two Rayleigh numbers of 10000 and 25000. According to the literature [1, 2], we should see different numbers (2 and 3 vortices at Ra=10000 and 25000, respectively) of vortices in the fluid at these two Rayleigh numbers. Note that in this example, we change the Ra number by changing the thermal conduction coefficient of the fluid.

Reviewer: ```suggestion
In this example, we simulate the Rayleigh-Benard convection problem at two Rayleigh numbers of 10000 and 25000. According to the literature [1, 2], we should see different numbers (2 and 3 vortices at Ra=10000 and 25000, respectively) of vortices in the fluid at these two Rayleigh numbers. Note that in this example, we change the Rayleigh number by changing the thermal conduction coefficient of the fluid.
```

.. note:: 

>     All the four boundary conditions are ``noslip``, and an external 

Reviewer 2: ```suggestion
    All four boundary conditions are ``noslip``, and an external 
```

    gravity field of :math:`-10` is applied in the :math:`y` direction.

--------------
Parameter file
--------------

Time integration is handled by a 1st order backward differentiation scheme 
`(bdf1)`, for a :math:`10000` s simulation time with an initial 
time step of :math:`0.01` second.

.. note::   

>     This example uses an adaptive time-stepping method, where the 

Reviewer: Maybe it would be interesting to explain why an adaptive time-stepping method is used.

Author: This has been explained in incompressible flow examples.

Reviewer: Yes you're right, i didn't see it. I would maybe add the reference to the other example in that case.

```suggestion
    As explained in `2D Flow past a backward-facing step <https://lethe-cfd.github.io/lethe/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.html>`_, this example uses an adaptive time-stepping method, where the 
```

    time-step is modified during the simulation to keep the maximum value of the CFL condition below a given threshold (0.5 here). Using ``output control = time``, and ``output time = 25`` the simulation results are written every 25 s.

.. note::   
    Note that the heating process is slow, and the velocity magnitudes are small inside the fluid. Hence, we expect large time-steps and a long simulation.

.. code-block:: text

    # --------------------------------------------------
    # Simulation Control
    #---------------------------------------------------
    subsection simulation control
        set method                          = bdf1
        set time end                        = 10000
        set time step                       = 0.01
        set adapt                           = true
        set max cfl                         = 0.5
        set stop tolerance                  = 1e-5
        set adaptative time step scaling    = 1.3

>         set number mesh adapt               = 0

Reviewer: ```suggestion
```
it is the default parameter

        set output name                     = rayleigh-benard_convection
        set output control                  = time
        set output time                     = 25
        set output path                     = ./output/     
    end

> The ``multiphysics`` subsection enables to turn on `(true)` and off `(false)` the physics of interest. Here ``heat transfer``, ``buoyancy force``, and ``fluid dynamics`` are chosen.

Reviewer 2: ```suggestion
The ``multiphysics`` subsection enables to turn on ``true`` and off ``false`` the physics of interest. Here ``heat transfer``, ``buoyancy force``, and ``fluid dynamics`` are chosen.
```

.. code-block:: text

    #---------------------------------------------------
    # Multiphysics
    #---------------------------------------------------
    subsection multiphysics
        set buoyancy force      = true
        set heat transfer       = true
        set fluid dynamics      = true
    end 
    

> The ``source term`` subsection defines gravitational acceleration. The value of the gravitational acceleration in this example is selected to satisfy the desired values of Ra number.

Reviewer: At line 53, it is stated that the Ra is set by changing the value of k. Here, one might think that it is set by changing the gravity magnitude. Maybe something like `` The gravity magnitude is set to -10 for both simulations. Thus, the Rayleigh number is controlled only by the thermal conduction coefficient for this example. `` could be added at line 53.

```suggestion
The ``source term`` subsection defines gravitational acceleration. The value of the gravitational acceleration in this example is set to -10.
```

Author: I've explained this issue in line 123 in detail, but thank you for your suggestion, this sentence is added to line 53 :)

.. code-block:: text
    
    #---------------------------------------------------
    # Source term
    #---------------------------------------------------
    subsection source term
        set enable                      = true
        subsection xyz
            set Function expression     = 0 ; -10 ; 0
        end
    end

> The ``physical properties`` subsection defines the physical properties of the fluid. Since we simulate the Rayleigh-Benard convection at two Rayleigh numbers (Ra=10000 and 25000), we use different thermal conductivities to reach mentioned Rayleigh numbers. We change the thermal conductivity of the fluid in the two simulations. Note that any other physical property (that is present in the Rayleigh number equation defined above) can be used instead of thermal conductivity. Both thermal conductivity values (:math:`k=0.15625` for :math:`Ra=10000`, and :math:`k=0.15625` for :math:`Ra=25000`) are added to the parameter handler file. However, only one of them should be uncommented for each simulation.

Reviewer 2: ```suggestion
The ``physical properties`` subsection defines the physical properties of the fluid. Since we simulate the Rayleigh-Benard convection at two Rayleigh numbers (Ra=10000 and 25000), we use different thermal conductivities to each mentioned Rayleigh numbers. We change the thermal conductivity of the fluid in the two simulations. Note that any other physical property (that is present in the Rayleigh number equation defined above) can be used instead of thermal conductivity. Both thermal conductivity values (:math:`k=0.15625` for :math:`Ra=10000`, and :math:`k=0.15625` for :math:`Ra=25000`) are added to the parameter handler file. However, only one of them should be uncommented for each simulation.
```

Reviewer 2: Should there rather be 2 parameter files? I know it's only a line, but for other examples there were two parameter files when running for two different viscosities, for example.

> The ``physical properties`` subsection defines the physical properties of the fluid. Since we simulate the Rayleigh-Benard convection at two Rayleigh numbers (Ra=10000 and 25000), we use different thermal conductivities to reach mentioned Rayleigh numbers. We change the thermal conductivity of the fluid in the two simulations. Note that any other physical property (that is present in the Rayleigh number equation defined above) can be used instead of thermal conductivity. Both thermal conductivity values (:math:`k=0.15625` for :math:`Ra=10000`, and :math:`k=0.15625` for :math:`Ra=25000`) are added to the parameter handler file. However, only one of them should be uncommented for each simulation.

Reviewer: ```suggestion
The ``physical properties`` subsection defines the physical properties of the fluid. Since we simulate the Rayleigh-Bénard convection at two Rayleigh numbers (Ra=10000 and 25000), we use different thermal conductivities to reach mentioned Rayleigh numbers. We change the thermal conductivity of the fluid in the two simulations. Note that any other physical property (that is present in the Rayleigh number equation defined above) can be used instead of thermal conductivity. Both thermal conductivity values (:math:`k=0.15625` for :math:`Ra=10000`, and :math:`k=0.15625` for :math:`Ra=25000`) are added to the parameter handler file. However, only one of them should be uncommented for each simulation.
```

.. code-block:: text

    #---------------------------------------------------
    # Physical Properties
    #---------------------------------------------------
    subsection physical properties
        set number of fluids            = 1
        subsection fluid 0
            set density                 = 100
            set kinematic viscosity     = 0.01
            set thermal expansion       = 0.0002
            set thermal conductivity    = 0.15625	# for Ra = 10000
            #set thermal conductivity   = 0.0625	# for Ra = 25000
            set specific heat           = 100
        end
    end

---------------------------
Running the simulation
---------------------------

Call the gls_navier_stokes_2d by invoking:  

``mpirun -np 8 gls_navier_stokes_2d rayleigh_benard_convection.prm``

to run the simulation using eight CPU cores. Feel free to use more.

> .. warning:: 

Reviewer 2: Nice addition to mention the approx. time on a number of cores !

    Make sure to compile lethe in `Release` mode and 
    run in parallel using mpirun. This simulation takes
    :math:`\approx` 20 minutes on 8 processes.

-------
Results
-------

> 

Reviewer: Is there an experimental/analytical result that we can use for comparison?

Reviewer 3: Good question, that would be very relevant!

Author: The only analysis that I could perform was the comparison of the number of vortices at different Rayleigh numbers. The behavior of flow in Rayleigh-Bénard convection is extremely unstable which causes bifurcation of the flow. This makes it very hard to compare other parameters with other experiments/simulations.

The following animation shows the results of this simulation:

.. raw:: html

    <iframe width="560" height="315" src="https://www.youtube.com/embed/tEg5M-wiCp8" frameborder="0" allowfullscreen></iframe>

> Note that at Ra=10000, two vortices exist in the fluid, while an extra (relatively small) vortex appear near the right wall. The velocity magnitude in the vortices is larger at smaller Rayleigh number.

Reviewer 2: ```suggestion
Note that at Ra=10000, two vortices exist in the fluid, while an extra (relatively small) vortex appears near the right wall. The velocity magnitude in the vortices is larger at smaller Rayleigh number.
```
